_posts/2020-02-17-eda-architecture.md

```diff
@@ -0,0 +1,181 @@
+
+Events are precisely defined messages that streams on a message bus, and each logical service (implementing one such business service as explained above) subscribes to the events that are relevant to it, without needing any knowledge about what produced them and how. They also push their own events on the bus, without caring about what consumes them.
+
+In that way, we completely decouple the services between each other and the message broker running the message bus provides us with load balancing, distribution and high availability for free. For that purpose, we chose [RabbitMQ](https://www.rabbitmq.com/) as message broker for its reliability record and ease of use.
+
+Now, the messages *are* the API, the only business and technical reference.
+
+After much thinking and experiments, we came with these core design principles that are very important, and after 4 months of production use, we are very glad we complied with them from the start:
+
+### Business Services and Data Processing Services
+
+We make a distinction between Business Services and Data Processing Services (a.k.a. utility services) to cleanly separate business logic from data processing complexity.
+
+DP Services are expected to be  **pure, stateless, services** that provide some kind of algorithmic data processing (computations, transformations, …). Moreover, they are also context free: they should not depend on business knowledge, assumptions or external data sources. All they need to do their processing must be in the command itself. They should not have to query a tier to get more data. DP services are kind of *universal* libraries and can even be provided by tiers.
+
+Business Services implement the customers' workflows and only focus on business rules and requirements to orchestrate and implement the value addition upon our customers' audio and data. They make use of the DP services as a library for that. There are very specific to us: you would never want to externalize your business services.
+
+Business Services persist the data they produce and are their unique trusted source of truth.
+
+Business Services build and maintain their own customer configuration from events on the bus.
+
+All services must be idempotent, that is, if they receive twice the exactly same message, they must behave identically and produce the same outputs.
+
+### Commands, Results and Events
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages emitted by business services and telling the bus what happened.
```

Comment:
I prefer "publish" instead of "emit"

Reply:
I don't know about the implementation, but emitting or publishing seems indeed different to me. You publish somewhere (so you know about the stuff you publish to), but you emit nowhere (stuff is listening for your emission, and you don't know about this stuff).

Reply:
"Publish" is the dedicated name for a message that is produced by an application (see https://www.rabbitmq.com/publishers.html) and the publish-subscribe pattern specifically use the term _publish_ (see https://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html)

I would say that "emit" is generic: a message is _emitted_ by an application. In contrast, a command is _sent_ and an event is _published_. Hence, using "publish" here seems more appropriate to me.